pkg/operator/ceph/csi/ceph_connection.go

```diff
@@ -0,0 +1,95 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package csi
+
+import (
+	"github.com/pkg/errors"
+
+	cephv1 "github.com/rook/rook/pkg/apis/ceph.rook.io/v1"
+	cephclient "github.com/rook/rook/pkg/daemon/ceph/client"
+
+	csiopv1a1 "github.com/ceph/ceph-csi-operator/api/v1alpha1"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func CreateUpdateCephConnection(c client.Client, clusterInfo *cephclient.ClusterInfo, clusterSpec *cephv1.ClusterSpec) error {
+
+	logger.Info("Creating ceph connection CR %q in namespace %q", clusterInfo.NamespacedName().Name, clusterInfo.NamespacedName().Namespace)
+	csiCephConnection := &csiopv1a1.CephConnection{}
+
+	csiCephConnection.Name = clusterInfo.NamespacedName().Name
+	csiCephConnection.Namespace = clusterInfo.NamespacedName().Namespace
+
+	spec, err := generateCephConnSpec(c, clusterInfo, &csiCephConnection.Spec, clusterSpec)
+	if err != nil {
+		return errors.Wrapf(err, "failed to set ceph connection CR %s in namespace %s", csiCephConnection.Name, clusterInfo.Namespace)
+	}
+	csiCephConnection.Spec = spec
+
+	err = clusterInfo.OwnerInfo.SetOwnerReference(csiCephConnection)
+	if err != nil {
+		return errors.Wrapf(err, "failed to set owner reference to ceph connection CR %q", csiCephConnection.Name)
+	}
+
+	err = c.Get(clusterInfo.Context, types.NamespacedName{Name: csiCephConnection.Name, Namespace: csiCephConnection.Namespace}, csiCephConnection)
+	if err != nil {
+		if kerrors.IsNotFound(err) {
+			err = c.Create(clusterInfo.Context, csiCephConnection)
+			if err != nil {
+				return errors.Wrapf(err, "failed to create ceph connection CR %s", csiCephConnection.Name)
+			}
+		} else if kerrors.IsAlreadyExists(err) {
+			err = c.Update(clusterInfo.Context, csiCephConnection)
+			if err != nil {
+				return errors.Wrapf(err, "failed to update ceph connection CR %s", csiCephConnection.Name)
+			}
+		}
+		return errors.Wrap(err, "failed to get ceph connection CR")
+	}
+
+	logger.Info("Successfully created ceph connection CR")
+	return nil
+}
+
+func generateCephConnSpec(c client.Client, clusterInfo *cephclient.ClusterInfo, csiClusterConnSpec *csiopv1a1.CephConnectionSpec, clusterSpec *cephv1.ClusterSpec) (csiopv1a1.CephConnectionSpec, error) {
+	if clusterSpec.CSI.ReadAffinity.Enabled {
+		csiClusterConnSpec.ReadAffinity.CrushLocationLabels = clusterSpec.CSI.ReadAffinity.CrushLocationLabels
+	}
+
+	cephRBDMirrorList := &cephv1.CephRBDMirrorList{}
+	err := c.List(clusterInfo.Context, cephRBDMirrorList, &client.ListOptions{Namespace: clusterInfo.Namespace})
+	if err != nil {
+		return *csiClusterConnSpec, err
```

Comment:
Try to return named errors where ever possible. 
```suggestion
		return *csiClusterConnSpec, fmt.Wrapf(err, "failed to list CephRBDMirror resource")
```